## Etapa 1. Descripción de los datos <a id='data_review'></a>

In [1]:
# Importar pandas
import pandas as pd
import numpy as np

In [2]:
# Leer el archivo y almacenarlo en df
df = pd.read_csv('/Users/luisdgr/Desktop/Proyecto 3 /music_project_en.csv')

In [3]:
# Obtener las 10 primeras filas de la tabla df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

In [4]:
# Obtener la información general sobre nuestros datos
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


Tenemos en resumen un dataframe con datos del tráfico que tiene la plataforma, dividido por usuario, por track, por artista y por género escuchado, por la ciudad donde lo reprodujeron, así como el día y la hora de cuando fue hecha la reproducción. A grandes rasgos, es un dataframe donde podemos ver la actividad por usuario.

Existen datos suficientes para dar conclusión y validar nuestra hipótesis, ya que los datos representan la actividad por día de los usuarios.

Definitivamente hay valores ausentes que no tienen mucho sentido en una inspección inicial, pues nos está diciendo que hay reproducciones de tracks no existentes, cuando todos los tracks tienen su título, además de artista y género. Hasta ahora duplicados no he encontrado con la inspección inicial y, respecto a los datatypes, hay algunos que podríamos pulir; como lo sería el tiempo a formato time, y quizá por comodidad podríamos cambiar los días de la semana por números, dependiendo las necesidades.

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>


In [5]:
# Muestra los nombres de las columnas
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


In [6]:
# Bucle en los encabezados poniendo todo en minúsculas
new_columns = []

for col in df.columns:
    new_name = col.strip()
    new_columns.append(new_name)

df.columns = new_columns


In [7]:
# Bucle en los encabezados eliminando los espacios
new_columns = []

for col in df.columns:
    new_name = col.strip().lower()
    new_columns.append(new_name)

df.columns = new_columns

In [8]:
# Cambiar el nombre de la columna "userid"

df.rename(columns={'userid': 'user_id'}, inplace=True)


In [9]:
# Comprobar el resultado: la lista de encabezados

print(df.columns)


Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


### Valores ausentes <a id='missing_values'></a>

In [10]:
# Calcular el número de valores ausentes
print(df.isna().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


In [11]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'

rep_columns = ['track', 'artist', 'genre']

for rep in rep_columns:

    df[rep] = df[rep].fillna('unknown')


In [12]:
# Contar valores ausentes

print(df.isna().sum())


user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


### Duplicados <a id='duplicates'></a>

In [13]:
# Contar duplicados explícitos

print(df.duplicated().sum())


3826


In [14]:
# Eliminar duplicados explícitos

df.drop_duplicates(inplace=True)
df.reset_index(drop=True, inplace=True)

print(df.head())


    user_id                        track            artist  genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile   rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg   rock   
2    20EC38            Funiculì funiculà       Mario Lanza    pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice   folk   
4  E2DC1FAE                  Soul People        Space Echo  dance   

          city      time        day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  


In [15]:
# Comprobar de nuevo si hay duplicados

print(df.duplicated().sum())


0


In [16]:
# Inspeccionar los nombres de géneros únicos

print(np.sort(df['genre'].unique()))


['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

In [17]:
# Función para reemplazar duplicados implícitos

def replace_wrong_genres(wrong_genres, correct_genre):

    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)


In [18]:
# Eliminar duplicados implícitos

replace_wrong_genres(['hip', 'hop', 'hip-hop'], "hiphop")


In [19]:
# Comprobación de duplicados implícitos

print(np.sort(df['genre'].unique()))


['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

### Observaciones <a id='data_preprocessing_conclusions'></a>

Al inicio noté que había 3,826 duplicados explícitos; y sabiendo que se trata de duplicados así únicamente había que eliminarlos con el método drop_duplicates(). De igual forma, algo que se nos mencionó en clase es que cuando se trate de eliminar filas, usualmente una buena práctica es hacer reset_index() para re-enumerar nuestras filas.

Ya que estaba hecho eso, al momento de analizar duplicados implícitos me quedé un poco más sorprendido, ya que había muchos valores únicos para la columna 'genre'. El ejercicio indica que trabajemos sobre lo relacionado a 'hip-hop' pero existen muchos más géneros y posibilidad de duplicados. Lo que sirvió mucho fue importar la librería numpy para poder organizar alfabéticamente los géneros y dar una lectura rápida a los valores e intentar ver si no existen valores similares de géneros. Hasta ahora creo que todo bien.

En sí la función es muy sencilla; lo único que hace es recibir una lista con valores incorrectos y el valor correcto como string; dentro de la función se verifica contra cada valor incorrecto de la lista y si procede, se reemplaza con el valor correcto.

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

In [20]:
print(df.head())

    user_id                        track            artist  genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile   rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg   rock   
2    20EC38            Funiculì funiculà       Mario Lanza    pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice   folk   
4  E2DC1FAE                  Soul People        Space Echo  dance   

          city      time        day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  


In [21]:
# Contar las canciones reproducidas en cada ciudad

df_count_by_city = df.groupby(by=['city'])['city'].value_counts()

print("Valores totales de actividad en lunes, miércoles y viernes para ambas ciudades: \n")
print(df_count_by_city)

Valores totales de actividad en lunes, miércoles y viernes para ambas ciudades: 

city
Shelbyville    18512
Springfield    42741
Name: count, dtype: int64


Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [22]:
# Calcular las canciones reproducidas en cada uno de los tres días

df_count_by_day = df.groupby(by=['city','day'])['city'].value_counts()

print("Valores separados de actividad en lunes, miércoles y viernes para ambas ciudades: \n")
print(df_count_by_day)

Valores separados de actividad en lunes, miércoles y viernes para ambas ciudades: 

city         day      
Shelbyville  Friday        5895
             Monday        5614
             Wednesday     7003
Springfield  Friday       15945
             Monday       15740
             Wednesday    11056
Name: count, dtype: int64



A primera vista, podemos observar que hay muchos más usuarios activos en Springfield en los 3 días que evaluamos, lunes, miércoles y viernes. Se podría deducir que hay más actividad en Springfield si omitiéramos el dato de cuántos habitantes tiene cada ciudad realmente. Haciendo otro paso de investigación podríamos tener un ratio y conocer qué porcentaje de cada población total tiene actividad dentro de la plataforma. Para fines del proyecto, nos quedamos con que hay más actividad en Springfield.

Por otro lado, podemos observar que en Springfield hay mayor actividad los viernes y en Shelbyville hay mayor actividad los miércoles. No obstante, la diferencia entre la actividad máxima (viernes) de Springfield vs la siguiente en la lista (lunes con 15,740), es una diferencia pequeña. Mientras que en Shelbyville hay una gran diferencia entre la mayor actividad (miércoles) vs la que le sigue en posición (viernes con 5,895).

Podríamos ir generando una hipótesis más donde diremos que en Springfield se podría considerar el lunes y viernes como los mejores días para hacer lanzamientos o novedades dentro de la plataforma de música, mientras que en Shelbyville tendría que tener mayor enfoque en el día miércoles para realizar este tipo de dinámicas.

In [ ]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(day, city):
   
    df_filtered = df[df['day'] == day]
  
    df_filtered = df_filtered[df_filtered['city'] == city]
    
    user_count = df_filtered['user_id'].count()
    
    # Devuelve el número de valores de la columna 'user_id'
    return user_count

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [24]:
# El número de canciones reproducidas en Springfield el lunes
number_tracks('Monday','Springfield')

np.int64(15740)

In [25]:
# El número de canciones reproducidas en Shelbyville el lunes
number_tracks('Monday','Shelbyville')

np.int64(5614)

In [26]:
# El número de canciones reproducidas en Springfield el miércoles
number_tracks('Wednesday','Springfield')

np.int64(11056)

In [27]:
# El número de canciones reproducidas en Shelbyville el miércoles
number_tracks('Wednesday','Shelbyville')

np.int64(7003)

In [28]:
# El número de canciones reproducidas en Springfield el viernes
number_tracks('Friday','Springfield')

np.int64(15945)

In [29]:
# El número de canciones reproducidas en Shelbyville el viernes
number_tracks('Friday','Shelbyville')

np.int64(5895)

**Conclusiones**

Tomando en cuenta que la hipótesis al inicio del proyecto fue:

"La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad."

Podemos afirmar que se cumple la hipótesis. Sin embargo, tiene unas cuantas excepciones que sí valdría la pena exponer.

Para el caso de los usuarios de la ciudad de Shelbyville, podemos observar que sí existe un día de la semana en particular donde se dispara la actividad, siendo este el día miércoles. Ahí se nota una clara diferencia pero, los otros días evaluados, están bastante cercanos en cuanto a conteos (cerca de 5,700)

En el caso de Springfield, el día más alto fue el viernes, pero aún así no estuvo muy lejano del anterior, que fue el lunes.

Retomando el punto inicial, "la actividad de los usuarios difiere según el día de la semana y dependiendo la ciudad", sí, existen diferencias marcadas.

# Conclusiones <a id='end'></a>

- 1 - La actividad de Springfield es mayor que la de Shelbyville. Puede ser por tema de población así que queda abierto ese tema.
- 2 - En Shelbyville, el día con mayor actividad es el miércoles, con una diferencia considerable vs el resto de días.
- 3 - En Springfield, hay dos días de mayor actividad porque son similares en sus conteos, siendo el viernes y el lunes. 
- 4 - Además de tener esos dos días de actividad alta en Springfield, la diferencia vs su conteo del día miércoles también es considerable.